### PR TITLE
feat: endpoint reachability plugin

### PR DESCRIPTION
**Pull request by 5war00p**
**Endpoint Reachability Plugin** 
- Brief Description of Changes
Added new plugin that checks and validates whether an endpoint (url) is reachable or unreachable.

**Motivation:** (optional)
- Why this change is necessary or beneficial
Its a preventive check to not do unnecessary actions without a proper input

**Related Issues:** (optional)
- #512

### CODE REVIEW

**Comment by roh26it** [comment]
Thanks for the PR @5war00p!

Does this only check the entire content of the request to assume that it contains a URL?

We also have another plugin - https://github.com/Portkey-AI/gateway/blob/main/plugins/default/validUrls.ts

Is that conflicting?

**Comment by 5war00p** [comment]
This just does almost same, but instead of GET request I'm using HEAD request also we as we need to validate endpoint is reaching or not just header is fine no need validate dns separately.

**Review by narengogi — COMMENTED** [review]
Looks good, just that this assumes that the response only contains a URL. in comparison,
https://github.com/Portkey-AI/gateway/blob/main/plugins/default/validUrls.ts feels more complete, I think it's a good idea to make a HEAD request to check the validity of a URL, you can perhaps add it as a fallback in the code of the existing plugin. That would be a 2 line change and more robust

**Comment by 5war00p** [comment]
To your point, `validUrl` ideally should be a regexp check thats what name speaks about. Where as `endpointReachability` make sure it reaches. If we really don't need 2 separate then we can merge, but validUrl can be used to validate a url input but if we make sure its a reachable one then merging both will be good idea.

File: plugins/default/endpointReachability.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  _parameters,
+  eventType,
+  _options
+) => {
+  let error = null;
+  let verdict = false;
+  let data = null;
+
+  try {
+    let text = getText(context, eventType);
```
Comment (b4s36t4) [inline plugins/default/endpointReachability.ts:27]: My suggestion here is use regex to get all the URLs available from text and validate them using `URL` or some other way. 

With this we can have config weather to run plugin for all the endpoints or first endpoint found within the text.

What do you think? @5war00p

**Review by b4s36t4 — CHANGES_REQUESTED** [review]

File: plugins/default/endpointReachability.ts (hunk 1 above)
Comment (5war00p) [reply plugins/default/endpointReachability.ts:27]: Makes sense if text might contain >1 url, however this issue is to build a plugin for endpoint reachability right? and we should use this util func in another which extracts urls and check each url using this plugin. Correct me if I went off.

Comment (b4s36t4) [reply plugins/default/endpointReachability.ts:27]: Your thought makes sense, but this PR is a plugin PR which I think doesn't make sense to use it as utility function. I think we should focus on keep this as individual plugin. keeping that I prefer we should go for regex match. 

As an enhancement I think we shouldn't just blindly hit the URL, we should ask user which type of method they're thinking of ex: `GET` and `POST` and let plugin do `OPTIONS` request and make sure the endpoint(s) support the user requested methods.

let me know if this makes sense to you.

Comment (5war00p) [reply plugins/default/endpointReachability.ts:27]: I think if thats the case we should make it as a new separate plugin.

**Comment by narengogi** [comment]
> To your point, `validUrl` ideally should be a regexp check thats what name speaks about. Where as `endpointReachability` make sure it reaches. If we really don't need 2 separate then we can merge, but validUrl can be used to validate a url input but if we make sure its a reachable one then merging both will be good idea.

sorry for the late reply @5war00p , as you might have noticed during your testing, the plugins act on the request and response body, so your plugin assumes it recieves only the URL in the body. Hence the plugin is not of much use.

But it's a cool implementation like I said, but for now this cannot be merged

**Comment by 5war00p** [comment]
@narengogi  In that case we can do as @b4s36t4 suggested, extract links and verify endpoint reachability, but question is do we need to do that for all links in found in entire text and how we do we determine that?

**Comment by narengogi** [comment]
@5war00p you can use regex to extract the URLs

**Comment by VisargD** [comment]
Hey folks! Are we moving forward with this implementation?

**Comment by VisargD** [comment]
Hey! Closing this PR because there has been no activity since November end. Please reopen it if you are still working on it.

**Pull request closed without merging**